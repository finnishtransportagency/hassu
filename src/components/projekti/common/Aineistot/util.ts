import { SelectOption } from "@components/form/Select";
import { Aineisto, AineistoInput, AineistoInputNew, AineistoNew, AineistoTila, VelhoAineisto } from "@services/api";
import { uuid } from "common/util/uuid";
import { AineistoKategoria, aineistoKategoriat, kategorisoimattomatId } from "hassu-common/aineistoKategoriat";
import find from "lodash/find";
import { Translate } from "next-translate";
import { Key } from "react";
import { FieldArrayWithId } from "react-hook-form";

export interface AineistotKategorioittain {
  [kategoriaId: string]: AineistoInput[];
}

export interface AineistoNahtavillaTableFormValuesInterface {
  aineistoNahtavilla: AineistotKategorioittain;
  poistetutAineistoNahtavilla: AineistoInput[];
}

export type FormAineisto = FieldArrayWithId<AineistoNahtavillaTableFormValuesInterface, `aineistoNahtavilla.${string}`, "id"> &
  Pick<Aineisto, "tila" | "tuotu" | "tiedosto">;

export type FormAineistoNew = AineistoInputNew & Pick<AineistoNew, "tuotu" | "tiedosto">;

export function getInitialExpandedAineisto(aineistot: AineistotKategorioittain | { [kategoriaId: string]: AineistoInputNew[] }): Key[] {
  const keyArray = [];
  const hasKategorisoimattomatAineisto = !!aineistot?.[kategorisoimattomatId]?.length;
  if (hasKategorisoimattomatAineisto) {
    keyArray.push(kategorisoimattomatId);
  }
  return keyArray;
}

export function getAllOptionsForKategoriat({
  kategoriat,
  ylakategoriaNimi,
  t,
}: {
  kategoriat: AineistoKategoria[];
  ylakategoriaNimi?: string;
  t: Translate;
}) {
  const ylakategoriaPrefix = ylakategoriaNimi ? `${ylakategoriaNimi} - ` : "";
  const kategoriaIds: SelectOption[] = [];
  kategoriat.forEach((kategoria) => {
    kategoriaIds.push({
      label: ylakategoriaPrefix + t(`aineisto-kategoria-nimi.${kategoria.id}`),
      value: kategoria.id,
    });
    if (kategoria.alaKategoriat) {
      kategoriaIds.push(
        ...getAllOptionsForKategoriat({
          kategoriat: kategoria.alaKategoriat,
          ylakategoriaNimi: ylakategoriaPrefix + t(`aineisto-kategoria-nimi.${kategoria.id}`),
          t,
        })
      );
    }
  });
  return kategoriaIds;
}

export function findKategoriaForVelhoAineisto(valitutVelhoAineistot: VelhoAineisto[]): AineistoInput[] {
  return valitutVelhoAineistot.map<AineistoInput>((velhoAineisto) => ({
    dokumenttiOid: velhoAineisto.oid,
    nimi: velhoAineisto.tiedosto,
    kategoriaId: aineistoKategoriat.findKategoria(velhoAineisto.kuvaus, velhoAineisto.tiedosto)?.id,
    tila: AineistoTila.ODOTTAA_TUONTIA,
    uuid: uuid.v4(),
  }));
}

export function findKategoriaForVelhoAineistoNew(valitutVelhoAineistot: VelhoAineisto[]): AineistoInputNew[] {
  return valitutVelhoAineistot.map<AineistoInputNew>((velhoAineisto) => ({
    dokumenttiOid: velhoAineisto.oid,
    nimi: velhoAineisto.tiedosto,
    kategoriaId: aineistoKategoriat.findKategoria(velhoAineisto.kuvaus, velhoAineisto.tiedosto)?.id,
    uuid: uuid.v4(),
  }));
}

export function combineOldAndNewAineistoWithCategories({
  oldAineisto,
  newAineisto,
}: {
  oldAineisto: AineistotKategorioittain;
  newAineisto: AineistoInput[];
}) {
  return newAineisto.reduce((combinedNewAndOld, velhoAineisto) => {
    if (!find(Object.values(combinedNewAndOld).flat(), { dokumenttiOid: velhoAineisto.dokumenttiOid })) {
      if (!velhoAineisto.kategoriaId && !combinedNewAndOld[kategorisoimattomatId]) {
        combinedNewAndOld[kategorisoimattomatId] = [];
      }
      if (velhoAineisto.kategoriaId && !combinedNewAndOld[velhoAineisto.kategoriaId]) {
        combinedNewAndOld[velhoAineisto.kategoriaId] = [];
      }
      const kategorianAineistot = combinedNewAndOld[velhoAineisto.kategoriaId || kategorisoimattomatId];
      kategorianAineistot.push({ ...velhoAineisto, jarjestys: kategorianAineistot.length });
    }
    return combinedNewAndOld;
  }, oldAineisto || {});
}

export function combineOldAndNewAineisto({ oldAineisto, newAineisto }: { oldAineisto?: AineistoInput[]; newAineisto: AineistoInput[] }) {
  return newAineisto.reduce((combinedNewAndOld, velhoAineisto) => {
    if (!find(combinedNewAndOld, { dokumenttiOid: velhoAineisto.dokumenttiOid })) {
      combinedNewAndOld.push({ ...velhoAineisto, jarjestys: combinedNewAndOld.length });
    }
    return combinedNewAndOld;
  }, oldAineisto || []);
}

export function adaptVelhoAineistoToAineistoInput(velhoAineisto: VelhoAineisto): AineistoInput {
  return {
    dokumenttiOid: velhoAineisto.oid,
    nimi: velhoAineisto.tiedosto,
    tila: AineistoTila.ODOTTAA_TUONTIA,
    uuid: uuid.v4(),
  };
}
