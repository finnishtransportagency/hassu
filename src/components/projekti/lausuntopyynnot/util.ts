import { LadattuTiedosto, LadattuTiedostoInput, LadattuTiedostoTila } from "@services/api";
import { LadattuTiedostoInputWithTuotu, SplitLadattuTiedostoInput } from "./types";
import find from "lodash/find";

export const handleLadattuTiedostoArrayForDefaultValues = (
  aineistot: LadattuTiedosto[] | null | undefined,
  addDefaultJarjestys: boolean
): SplitLadattuTiedostoInput => {
  const initialSplitLadattuTiedostoInput: SplitLadattuTiedostoInput = {
    poistettu: [] as LadattuTiedostoInputWithTuotu[],
    lisatty: [] as LadattuTiedostoInputWithTuotu[],
  };

  return (
    aineistot
      ?.map(mapLadattuTiedostoToInput)
      .reduce(reduceToLisatytJaPoistetutLadattuTiedosto(addDefaultJarjestys), initialSplitLadattuTiedostoInput) ||
    initialSplitLadattuTiedostoInput
  );
};

export const reduceToLisatytJaPoistetutLadattuTiedosto =
  (addDefaultJarjestys: boolean) =>
  (
    acc: { poistettu: LadattuTiedostoInput[]; lisatty: LadattuTiedostoInput[] },
    tiedosto: LadattuTiedostoInput
  ): { poistettu: LadattuTiedostoInput[]; lisatty: LadattuTiedostoInput[] } => {
    if (tiedosto.tila === LadattuTiedostoTila.ODOTTAA_POISTOA || tiedosto.tila === LadattuTiedostoTila.POISTETTU) {
      acc.poistettu.push(tiedosto);
    } else {
      const aineistoToAdd =
        addDefaultJarjestys && typeof tiedosto.jarjestys !== "number" ? { ...tiedosto, jarjestys: acc.lisatty.length } : tiedosto;
      acc.lisatty.push(aineistoToAdd);
    }
    return acc;
  };

const mapLadattuTiedostoToInput = ({ nimi, jarjestys, tila, tiedosto, tuotu }: LadattuTiedosto): LadattuTiedostoInputWithTuotu => ({
  tiedosto,
  jarjestys,
  nimi,
  tila,
  tuotu,
});

export function combineOldAndNewLadattuTiedosto({
  oldTiedostot,
  oldPoistetut,
  newTiedostot,
}: {
  oldTiedostot?: LadattuTiedostoInputWithTuotu[];
  oldPoistetut?: LadattuTiedostoInputWithTuotu[];
  newTiedostot: LadattuTiedostoInputWithTuotu[];
}) {
  return newTiedostot.reduce<{ lisatyt: LadattuTiedostoInputWithTuotu[]; poistetut: LadattuTiedostoInputWithTuotu[] }>(
    (acc, tiedosto) => {
      if (!find(acc.lisatyt, { nimi: tiedosto.nimi })) {
        acc.lisatyt.push({ ...tiedosto, jarjestys: acc.lisatyt.length });
      }
      acc.poistetut = acc.poistetut.filter((poistettu) => poistettu.nimi !== tiedosto.nimi);
      return acc;
    },
    { lisatyt: oldTiedostot || [], poistetut: oldPoistetut || [] }
  );
}
