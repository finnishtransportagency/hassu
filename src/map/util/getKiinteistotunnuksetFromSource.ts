import VectorSource from "ol/source/Vector";

export const getKiinteistotunnuksetFromSource = (geoJsonSource: VectorSource) => {
  return geoJsonSource.getFeatures().reduce<Set<string>>((set, feat) => {
    const kiinteistotunnus = feat.getProperties().kiinteistotunnus;
    if (kiinteistotunnus) {
      set.add(kiinteistotunnus);
    }
    return set;
  }, new Set());
};
