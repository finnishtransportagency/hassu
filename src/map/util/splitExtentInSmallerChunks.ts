import { ProjectionLike, transform } from "ol/proj";
import { Extent } from "ol/extent";

export type SplitExtentInSmallerChunksOptions = {
  extent: Extent;
  chunkSquareSideLength: number;
  dataProj: ProjectionLike;
  viewProj: ProjectionLike;
};

export function splitExtentInSmallerChunks({ extent, chunkSquareSideLength, dataProj, viewProj }: SplitExtentInSmallerChunksOptions) {
  const [minX, minY] = coordViewToData([extent[0], extent[1]], dataProj, viewProj);
  const [maxX, maxY] = coordViewToData([extent[2], extent[3]], dataProj, viewProj);

  const xRanges = splitRange(minX, maxX, chunkSquareSideLength);
  const yRanges = splitRange(minY, maxY, chunkSquareSideLength);

  const allExtents = xRanges.reduce<Extent[]>((extents1, xRange) => {
    extents1.push(
      ...yRanges.reduce<Extent[]>((extents2, yRange) => {
        extents2.push([xRange[0], yRange[0], xRange[1], yRange[1]]);
        return extents2;
      }, [])
    );
    return extents1;
  }, []);
  return allExtents;
}
function coordViewToData(coord: number[], dataProj: ProjectionLike, viewProj: ProjectionLike) {
  if (viewProj == dataProj) {
    return coord;
  }
  return transform(coord, viewProj, dataProj);
}
function splitRange(rangeStart: number, rangeEnd: number, intervalSize: number) {
  const rangeDifference = rangeEnd - rangeStart;
  const numIntervals = rangeDifference / intervalSize;
  const intervals = [];

  for (let i = 0; i < numIntervals; i++) {
    const start = rangeStart + i * intervalSize;
    const end = Math.min(start + intervalSize, rangeEnd);
    intervals.push([start, end]);
  }
  return intervals;
}
