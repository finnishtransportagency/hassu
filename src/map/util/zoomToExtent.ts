import { Extent, isEmpty } from "ol/extent";
import { fromExtent as polygonFromExtent } from "ol/geom/Polygon.js";
import Map from "ol/Map";

export function zoomToExtent(map: Map, extent?: Extent | null) {
  const view = map.getView();
  const extentToFit = extent ?? view.getProjection().getExtent();
  if (!isEmpty(extentToFit)) {
    view.fitInternal(polygonFromExtent(extentToFit));
  }
}
