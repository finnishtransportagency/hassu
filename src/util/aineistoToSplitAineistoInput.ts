import { Aineisto, AineistoInput } from "@services/api";
import { mapAineistoToInput } from "./mapAineistoToInput";
import { reduceToLisatytJaPoistetut } from "./reduceToLisatytJaPoistetut";

type SplittedAineistoInput = { poistettu: AineistoInput[]; lisatty: AineistoInput[] };

const initialSplittedAineistoInput: SplittedAineistoInput = { poistettu: [], lisatty: [] };

export const aineistoToSplitAineistoInput = (aineistot?: Aineisto[] | null | undefined): SplittedAineistoInput =>
  aineistot?.map(mapAineistoToInput).reduce(reduceToLisatytJaPoistetut, initialSplittedAineistoInput) || initialSplittedAineistoInput;
