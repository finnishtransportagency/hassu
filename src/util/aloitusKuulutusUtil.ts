import dayjs from "dayjs";
import { SahkopostiVastaanottaja } from "../../common/graphql/apiModel";

export function examineKuulutusPaiva(givenKuulutusPaiva: string | null | undefined) {
  let published: boolean;
  let kuulutusPaiva: string | undefined;
  const date = givenKuulutusPaiva;
  if (date) {
    let parsedDate = dayjs(date);
    if (date.length == 10) {
      kuulutusPaiva = parsedDate.format("DD.MM.YYYY");
    } else {
      kuulutusPaiva = parsedDate.format("DD.MM.YYYY HH:mm");
    }
    published = parsedDate.isBefore(dayjs());
  } else {
    published = false;
    kuulutusPaiva = undefined;
  }
  return { kuulutusPaiva, published };
}

export function lahetysTila(vastaanottaja: Pick<SahkopostiVastaanottaja, "lahetysvirhe" | "lahetetty">) {
  if (vastaanottaja.lahetysvirhe) {
    return "Lähetysvirhe";
  }
  return vastaanottaja.lahetetty ? "Lähetetty" : "Ei lähetetty";
}

