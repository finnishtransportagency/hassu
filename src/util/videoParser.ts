const youtubeBaseURL = "https://www.youtube.com/embed/";
const vimeoBaseURL = "https://player.vimeo.com/video/";
export const parseVideoURL = (url: string): string | undefined => {
  if (!url) {
    return undefined;
  }
  if (url.includes("youtube")) {
    const youtubeID = url.split("v=")?.[1];
    return youtubeID ? youtubeBaseURL.concat(youtubeID) : undefined;
  } else if (url.includes("vimeo")) {
    const lastIndex = url.lastIndexOf("/");
    if (lastIndex < 0) {
      return undefined;
    }
    const vimeoID = url.substring(lastIndex + 1);
    if (isNaN(Number(vimeoID))) {
      return undefined;
    }
    return vimeoBaseURL.concat(vimeoID);
  }
  // not supported yet
  return undefined;
};
