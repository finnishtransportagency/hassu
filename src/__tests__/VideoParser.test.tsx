/**
 * @jest-environment jsdom
 */

import { parseVideoURL } from "../util/videoParser";

describe("VideoParser", () => {
  it("returns embedded youtube urls correctly", () => {
    const url = "https://www.youtube.com/watch?v=jNQXAC9IVRw";
    const embeddedURL = parseVideoURL(url);
    expect(embeddedURL).toMatchSnapshot();
  });

  it("accepts youtube urls with v query parameters only", () => {
    const url = "https://youtu.be/dQw4w9WgXcQ";
    const embeddedURL = parseVideoURL(url);
    expect(embeddedURL).toBeUndefined();
  });

  it("returns embedded vimeo urls correctly", () => {
    const url = "https://vimeo.com/76979871";
    const embeddedURL = parseVideoURL(url);
    expect(embeddedURL).toMatchSnapshot();
  });

  it("accepts vimeo urls with ID (number) as last part of path only", () => {
    const url = "https://player.vimeo.com/video/71994339?h=0173a63fdf";
    const embeddedURL = parseVideoURL(url);
    expect(embeddedURL).toBeUndefined();
  });

  it("returns values only for youtube or vimeo urls", () => {
    const url = "https://www.dailymotion.com/video/x3rdtfy";
    const embeddedURL = parseVideoURL(url);
    expect(embeddedURL).toBeUndefined();
  });
});
