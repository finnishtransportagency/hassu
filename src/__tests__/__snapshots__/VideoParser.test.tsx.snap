// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoParser returns embedded vimeo urls correctly 1`] = `"https://player.vimeo.com/video/76979871"`;

exports[`VideoParser returns embedded youtube urls correctly 1`] = `"https://www.youtube.com/embed/jNQXAC9IVRw"`;
